tools/depcheck/README.md

```diff
@@ -0,0 +1,176 @@
+]
+```
+
+...the internal graph that is built will still contain nodes for every Go package traversed in
+the repo, as well as the `api` and `auth` packages, but the sub-trees of these two packages
+will be "squashed" into `api` and `auth`. This means that while the child nodes of these two
+packages are deleted, the `edges` of those nodes to other sub-trees are preserved.
+
+For example:
+
+Given the following set of package dependencies (assume directed edges from parent to child, starting with `A` and `D`):
+
+```
+      A                D
+     / \              / \
+(A/B)   (A/C) -> (D/E)   (D/F)
+     \
+      (A/B/C)
+```
+
+Since the set of packages contained within repository `A` brings in a package `A/C` that depends on a package `D/E`, which is brought in by repository `D`,
+"squashing" these two sets of packages into their roots `A` and `D` results in the following:
+
+```
+      A -> D
+```
+
+[See the `depcheck` directory for filter examples](https://github.com/openshift/origin/tree/master/tools/depcheck/examples)
+
+### Excludes
+
+Just as it is possible to squash sub-packages into a parent package within a repository, it 
+is also possible to exclude entire directory sub-trees from a repository.
+
+Example:
+
+Given the following file...
+```
+$ cat excludes.json
+[
+    "github.com/openshift/origin/images",
+    "github.com/openshift/origin/pkg/build/builder"
+    "github.com/openshift/origin/cmd/service-catalog"
+]
+```
+
+...the internal graph that is built will all exclude sub-packages within the
+`images`, `builder`, and `service-catalog` paths.
+
+### Analyze
+
+The analyze command requires the following flags:
+  - `--root=`  The `Go` import path of the repository root (e.g. `github.com/foo/bar`)
+  - `--entry=` The (relative) path for each of the entry-point directories to be used to begin traversing the repository's directory structure. (e.g. `./pkg`)
+  - `--dep=`   The full `Go` import path of the vendored dependency whose dependencies we are analyzing against our own (e.g. `github.com/foo/bar/vendor/k8s.io/kubernetes`).
+ 
+Optionally, the following flags may be specified as well:
+  - `--filter=`   The path to the `filters.json` file containing package paths to "squash".
+  - `--exclude=`  The path to the `excludes.json` file containing package paths to exclude from the internal graph and analysis.
+  - `--openshift` Mutually exclusive with `--filter` and `--exclude`. Uses `openshift` specific repository information to generate an internal list of package paths to exclude and filter (requires that the `origin` repository exist within your `GOPATH`).
+
+Example:
+
+Invoke the `analyze` command using `openshift`-specific defaults, comparing dependencies against those of `kubernetes`:
+
+```bash
+$ depcheck analyze --root=github.com/openshift/origin --entry=cmd/... --entry=pkg/... --entry=tools/... --entry=test/... --openshift --dep=github.com/openshift/origin/vendor/k8s.io/kubernetes
```

Comment:
I created the tmp gopath exported `$GOPATH` and run this command insinde `tools/godepcheck` (because that's where the binary is) and in origin root and I got:
```
$ ./depcheck analyze --root=github.com/openshift/origin --entry=cmd/... --entry=pkg/... --entry=tools/... --entry=test/... --openshift --dep=github.com/openshift/origin/vendor/k8s.io/kubernetes
can't load package: package github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/apis/settings: code in directory /tmp/depcheck/src/github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/apis/settings expects import "github.com/kubernetes-incubator/service-catalog/pkg/apis/settings"
can't load package: package github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/apis/settings/v1alpha1: code in directory /tmp/depcheck/src/github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/apis/settings/v1alpha1 expects import "github.com/kubernetes-incubator/service-catalog/pkg/apis/settings/v1alpha1"
can't load package: package github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/registry/settings/podpreset: code in directory /tmp/depcheck/src/github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/registry/settings/podpreset expects import "github.com/kubernetes-incubator/service-catalog/pkg/registry/settings/podpreset"
Error: exit status 1
Usage:
 ...
exit status 1
```

Btw. with error don't print all the usage, unless I specify wrong flag, similarly how we do it with `oc`.

Reply:
Thanks, will look into this. The `cmd/service-catalog` path should have been ignored as part of --openshift defaults

Reply:
Yeah, but ensure there's an example how to include it, when service-catalog folks will want to use your tool as well. And most importantly that it also works for them.

Reply:
Although, you might sync with them if this requires additional eng effort, before actually doing it.

Reply:
This issue is addressed in the last commit. Although a user is able to --exclude a list of Go import paths, those are only excluded when building the graph, not when traversing the directory tree. The new patch makes it so that any errors encountered while traversing an "excluded" path are non-fatal.

> Yeah, but ensure there's an example how to include it, when service-catalog folks will want to use your tool as well. And most importantly that it also works for them.

If they wish to use this tool, they either have to run it against their own independent repo, or include all of their specific vendored deps in the openshift/origin vendor directory, as Go will not look in their own nested vendor tree for those deps.